### DISCUSSION TITLE
What is the best way to enable endpoint discovery declaratively for multi cluster

### DISCUSSION THREAD

**Discussion by moore-nathan** [istio/istio#53433] [Q&A]
From the [multi cluster docs](https://istio.io/latest/docs/setup/install/multicluster/multi-primary/#enable-endpoint-discovery) it shows that to enable discovery we use `istioctl`.
Is it possible to declaratively set these? If so what are the steps needed to do so?

**Comment by howardjohn** [comment]
There is some recent discussion here https://github.com/istio/istio/discussions/53369#discussioncomment-10811785 but not yet fully complete. The answer is basically yes but its not clearly documented anywhere.

**Reply by moore-nathan** [reply]
Ahh ok I see.
I also wanted to ask specifically about the `create-remote-secret` command. After running it I can see everything in there is quite similar to a kubeconfig and I can get elsewhere to declaratively create it. The only spot I do not know how to match is the users section and token that gets created. I don't see that mentioned in the other thread, but if its easier let me know if I should ask that on the other thread to keep things together

**Reply by howardjohn** [reply]
https://github.com/istio/istio/blob/1ad41e17ee31bbaf7acf131b61da894be8f22303/istioctl/pkg/multicluster/remote_secret.go#L632 is the logic. Not super trivial but basically its coming from the istio-reader service account.

Ultimately what you need is a way to authenticate to an identity with the right permission, a label to identify it, and I think the name of the key matters

**Reply by moore-nathan** [reply]
For endpoint discovery is the ServiceAccount approach the only way to accomplish this? Meaning for Istio multi cluster I need to have a secret created and that secret must have the token from another cluster's ServiceAccount?

**Reply by keithmattix** [reply]
Yes pretty much

**Comment by moore-nathan** [comment]
Thank you for your guys' comments, it very much helped understand what was needed.
Wound up accomplishing this using external-secrets after figuring out what all was needed.
The istio helm charts at https://istio-release.storage.googleapis.com/charts do not automatically create a `kubernetes.io/service-account-token` type secret so had to add that into templates. This is the same step that is done in that `createRemoteSecret` command.
Once have that secret created I was able to use a [PushSecret](https://external-secrets.io/latest/api/pushsecret/) to store the token, and then use an [ExternalSecret](https://external-secrets.io/latest/api/externalsecret/) with templating to generate the `istio-remote-secret-{{ clusterName }}` that would otherwise be created from the istioctl. The format which is in that secret is important so had to run the `createRemoteSecret` command without kubectl applying to get a reference of what it looks like.
Overall this works quite well, only manual necessities is have to create `ExternalSecret` for every cluster need to connect to, which means have to specify that beforehand. Feel this is what would be expected though as was not expecting any automatic cluster discovery.

**Reply by josemrs** [reply]
> manual necessities is

That is basically what I ended up with. In the remote cluster a simple secret for the SA is created using a template then pushed to SSM Paramters using ESO PushSecret. In the Primary, ESO pulls the secret. Then, I have a Python script in a pod to monitor the pulled secrets and query EKS looking for cluster endpoints and CAs and finally create secrets with the expected content by Istio. 

The Primary is connecting and mutating the webhookconfiguration so I think the "secrets automation" works. I'm still struggling with the webhook in the remote not connecting to the primary, but that is another story.
